libcontainer/cgroups/systemd/user.go

```diff
@@ -0,0 +1,86 @@
+// +build linux
+
+package systemd
+
+import (
+	"bufio"
+	"bytes"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	systemdDbus "github.com/coreos/go-systemd/v22/dbus"
+	dbus "github.com/godbus/dbus/v5"
+	"github.com/opencontainers/runc/libcontainer/system"
+	"github.com/pkg/errors"
+)
+
+func newUserSystemdDbus() (*systemdDbus.Conn, error) {
+	addr, err := detectUserDbusSessionBusAddress()
+	if err != nil {
+		return nil, err
+	}
+	uid, err := detectUID()
+	if err != nil {
+		return nil, err
+	}
+
+	return systemdDbus.NewConnection(func() (*dbus.Conn, error) {
+		conn, err := dbus.Dial(addr)
+		if err != nil {
+			return nil, err
+		}
+		methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(uid))}
+		err = conn.Auth(methods)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+		if err = conn.Hello(); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		return conn, nil
+	})
+}
+
+// detectUID detects UID from $XDG_RUNTIME_DIR if running in userNS.
+// Otherwise returns os.Getuid() .
+func detectUID() (int, error) {
```

Comment:
I'm hoping you wouldn't need this... But looks like you still do for the external auth of the user D-Bus connection... Let me think about this one see if there's a better solution than this.

Having said that, see my comment below on requiring that the environment variables are properly set. So you could say that this is a hard requirement here. Looks like your code will fail if it doesn't match `/run/user/%d` which I think is a good thing.

Reply:
libsystemd (which is used by crun) seems extracting UID from the `"0::/user.slice/user-%d.slice/..."` string  in `/proc/self/cgroup`:
https://github.com/systemd/systemd/blob/6829d8ce6927c151e4391c6a5e69b76906fbd2be/src/libsystemd/sd-bus/sd-bus.c#L104
https://github.com/systemd/systemd/blob/ce4c73eb5d3c9a23628c450db0e25a529620adbe/src/basic/cgroup-util.c#L1361-L1395

I feel this seems less robust than `$XDG_RUNTIME_DIR`, but maybe we should do this?

Reply:
I'm :+1: for `path.Base(xdr)` proposed by @kolyshkin https://github.com/opencontainers/runc/pull/2281#discussion_r401778383

Reply:
Yet another way is to exec `busctl --user status` and grep `^UID=`

Reply:
`busctl --user status` ends up resolving it from `getsockopt(fd, SOL_SOCKET, SO_PEERCRED, ...)`, see [here](https://github.com/systemd/systemd/blob/v245/src/basic/socket-util.c#L735).

I believe you can do the same in Golang with `unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)` from the `"golang.org/x/sys/unix"` package. Does that work?

Reply:
> libsystemd (which is used by crun) seems extracting UID from the `"0::/user.slice/user-%d.slice/..."` string in `/proc/self/cgroup`:

Yes, I think that's possibly reasonable. It's still reimplementing C code in Golang but I think that's preferrable than some of the alternatives.

But see my message above about getting the real UID from a `SO_PEERCRED` call...

Reply:
> busctl --user status ends up resolving it from getsockopt(fd, SOL_SOCKET, SO_PEERCRED, ...), see here.

Seems not true for `OwnerUID`. `OwnerUID` seems extracted from cgroup.
https://github.com/systemd/systemd/blob/v245/src/libsystemd/sd-bus/bus-dump.c#L375-L377
https://github.com/systemd/systemd/blob/v245/src/libsystemd/sd-bus/bus-creds.c#L509-L526

Anyway, `SO_PEERCRED` isn't what we want here, because the kernel automatically translates the UID in the ucred struct.

Reply:
Updated PR to use `busctl --user status`. I think this is the most robust way. 

I guess eventually systemd may switch away from parsing cgroup and adopt more robust logic , but we won't be affected as long as we use `busctl`.

Reply:
Ah yes good point sorry my search was too shallow for it...

I think I just worry a bit about `busctl` not being available, it was first introduced as a debugging tool so I can see how distributions might not want to ship it or not ship it by default and it might go missing in the future...

Anyways, it's a fallback and we can iterate on it. Looks good.